### PR TITLE
riseupvpn experiment

### PR DESCRIPTION
**Pull request by cyBerta**
* Implements connectivity tests for riseupvpn including provider API and gateways.
* Tests the reachability of openvpn and obfs4 ports of each gateway by performing a tcp handshake.
* Includes integration tests

### CODE REVIEW

**Review by bassosimone — CHANGES_REQUESTED** [review]
Thank you so much for working in this topic! This review accounts for around 50% of the pull request and specifically does not cover the test suite, which I could not look into in detail.

As an overall comment, the implementation here looks good. Most of my comments, suggestions, and request pertain to code style, consistency, and fixes.

My recommendation is for you to apply the requested changes while I continue my work of code review and take a more in depth look at the tests, run the code locally, etc.

I think it's reasonable to assume this process will converge in the next two weeks.

I also think we need to have a broader conversation on the scope of the experiment, its design, and its aims. But I also feel like it's important that a design discussion is anchored to running code and that it makes sense to move this prototype forward for informing the discussion.

Thanks again, keep up the good work! :guitar:

**Comment by cyBerta** [comment]
Thanks for your review so far @bassosimone ! I'll adapt the PR accordingly to your comments.

File: experiment/riseupvpn/riseupvpn.go [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+
+// Config contains the riseupvpn experiment config.
+type Config struct {
+	urlgetter.Config
+}
+
+// TestKeys contains riseupvpn test keys.
+type TestKeys struct {
+	urlgetter.TestKeys
+	RiseupVPNApiFailure      *string             `json:"riseupvpn_api_failure"`
```
[the remark this answers is not in the document]
Comment (cyBerta) [reply experiment/riseupvpn/riseupvpn.go:65]: oh, it was inspired by tests for telegram, whatsapp, fbmanager. If Riseupvpn runs in the circumvention tools test suite together with tor and psiphon, would the scope still distinguishable?

File: experiment/riseupvpn/riseupvpn.go [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+		measurement.TestKeys = testkeys
+		return nil
+	}
+
+	ok := netx.CertPool.AppendCertsFromPEM([]byte(tk.HTTPResponseBody))
+	if !ok {
+		testkeys.RiseupVPNCACertStatus = false
+		testkeys.RiseupVPNApiStatus = "blocked"
+		errorValue := "invalid_ca"
+		testkeys.RiseupVPNApiFailure = &errorValue
```
[the remark this answers is not in the document]
Comment (cyBerta) [reply experiment/riseupvpn/riseupvpn.go:182]: agreed

File: experiment/riseupvpn/riseupvpn.go [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+	obfs4Endpoints := generateMultiInputs(gateways, "obfs4")
+
+	// measure openvpn in parallel
+	multi = urlgetter.Multi{Begin: time.Now(), Getter: m.Getter, Session: sess}
+	for entry := range multi.Collect(ctx, openvpnEndpoints, "riseupvpn", callbacks) {
+		testkeys.AddGatewayConnectTestKeys(entry, "openvpn")
+	}
+
+	// measure obfs4 in parallel
+	multi = urlgetter.Multi{Begin: time.Now(), Getter: m.Getter, Session: sess}
```
[the remark this answers is not in the document]
Comment (cyBerta) [reply experiment/riseupvpn/riseupvpn.go]: Thanks for the hint. I fixed that

File: experiment/riseupvpn/riseupvpn.go [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+
+//DecodeEIP3 decodes eip-service.json version 3
+func DecodeEIP3(body string) (*EipService, error) {
+	var eip EipService
+	err := json.Unmarshal([]byte(body), &eip)
+	if err != nil {
+		log.Error(err.Error())
+	}
+
+	return &eip, err
```
[the remark this answers is not in the document]
Comment (cyBerta) [reply experiment/riseupvpn/riseupvpn.go]: agreed. I took your suggestion

File: experiment/riseupvpn/riseupvpn.go [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+	ok := netx.CertPool.AppendCertsFromPEM([]byte(tk.HTTPResponseBody))
+	if !ok {
+		testkeys.RiseupVPNCACertStatus = false
+		testkeys.RiseupVPNApiStatus = "blocked"
+		errorValue := "invalid_ca"
+		testkeys.RiseupVPNApiFailure = &errorValue
+		return nil
+	}
+
+	urlgetter.RegisterExtensions(measurement)
```
[the remark this answers is not in the document]
Comment (cyBerta) [reply experiment/riseupvpn/riseupvpn.go]: Thanks, I fixed it

**Comment by cyBerta** [comment]
besides the open question about the RiseupVPN prefix in variables and the missing spec description of the error values I adapted the PR accordingly.

File: experiment/riseupvpn/riseupvpn.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (cyBerta) [reply experiment/riseupvpn/riseupvpn.go:182]: I added a PR: https://github.com/ooni/spec/pull/203

File: experiment/riseupvpn/riseupvpn.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (bassosimone) [reply experiment/riseupvpn/riseupvpn.go:65]: > oh, it was inspired by tests for telegram, whatsapp, fbmanager

Ah, my apologies, I didn't think about these specific keys in these experiments. I still have a feeling we don't need to add the `riseupvpn` prefix here. Going back, I'd not add any prefix to the IM experiment keys. We cannot change the past, but I think it's reasonable to use a more lean convention for this new experiment! 

> If Riseupvpn runs in the circumvention tools test suite together with tor and psiphon, would the scope still be distinguishable?

Yes: every experiment produces its own JSON document, therefore there is no ambiguity!

[the remark this answers is not in the document]
Comment (cyBerta) [reply experiment/riseupvpn/riseupvpn.go:65]: > Yes: every experiment produces its own JSON document, therefore there is no ambiguity!

Great! I'll remove the prefex then.

**Comment by cyBerta** [comment]
> Thank you for addressing my previous comments! I think there are two mains aspects of this diff that still need to be addressed, along with a bunch of minor nits. These aspects are:
> 
>     1. making sure we do not modify the global certificate pool (which should not have been exposed and which I'll make private soon)
> 
>     2. simplifying and possibly making more uniform the two tests in the test suite that deal with blocking of either "openvpn" or "obfs4" endpoints.
> 
> 
> Thanks again! +1

Thanks so much for your review again. 

I initially intended to set the CA certs only for the HTTP requests needed, as you described. Unfortunately I didn't (and still don't) see a possibility to set the generated cert pool to an urlgetter. I proposed a solution here: https://github.com/ooni/probe-engine/issues/1018#issuecomment-719922725. I would be interested if you see other already existing solutions to the problem.

File: experiment/riseupvpn/riseupvpn_test.go [hunk 6]
```diff
@@ -0,0 +1,480 @@
...
+	// - self censor random gateway
+	gateways := eipService.Gateways
+	if gateways == nil || len(gateways) == 0 {
+		t.SkipNow()
+	}
+
+	var selfcensoredGateways []string
+	for _, gateway := range gateways {
+		for _, transport := range gateway.Capabilities.Transport {
+			if transport.Type == "obfs4" {
```
[the remark this answers is not in the document]
Comment (cyBerta) [reply experiment/riseupvpn/riseupvpn_test.go:429]: You're right. Currently all gateways always support openvpn and it's always the first transport in the json array. But it's cleaner to filter for openvpn in TestIntegrationFailureOpenvpnGateway. :+1:

File: experiment/riseupvpn/riseupvpn_test.go [hunk 7]
```diff
@@ -0,0 +1,480 @@
...
+	}
+
+	if tk.ApiFailure == nil {
+		t.Fatal("ApiFailure should not be null")
+	}
+
+	cancel()
+}
+
+func TestIntegrationFailureOpenvpnGateway(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (cyBerta) [reply experiment/riseupvpn/riseupvpn_test.go]: fixed by implementing subtests

File: experiment/riseupvpn/riseupvpn_test.go [hunk 8]
```diff
@@ -0,0 +1,480 @@
...
+	// - fetch client cert and add to certpool
+	caFetchClient := &http.Client{
+		Timeout: time.Second * 30,
+	}
+
+	caCertResponse, err := caFetchClient.Get("https://black.riseup.net/ca.crt")
+	if err != nil {
+		t.SkipNow()
```
[the remark this answers is not in the document]
Comment (cyBerta) [reply experiment/riseupvpn/riseupvpn_test.go:273]: I'm skipping all failures that are related to the precondition to the actual testing like fetching the CA cert in order to gather all necessary gateway information.

File: experiment/riseupvpn/riseupvpn_test.go [hunk 9]
```diff
@@ -0,0 +1,480 @@
...
+		},
+	}
+
+	eipResponse, err := client.Get("https://api.black.riseup.net/3/config/eip-service.json")
+	if err != nil {
+		t.SkipNow()
+	}
+	defer eipResponse.Body.Close()
+
+	if eipResponse.StatusCode == http.StatusOK {
```
[the remark this answers is not in the document]
Comment (cyBerta) [reply experiment/riseupvpn/riseupvpn_test.go]: fixed

File: experiment/riseupvpn/riseupvpn_test.go [hunk 10]
```diff
@@ -0,0 +1,480 @@
...
+	}
+
+	caCertResponse, err := caFetchClient.Get("https://black.riseup.net/ca.crt")
+	if err != nil {
+		t.SkipNow()
+	}
+	defer caCertResponse.Body.Close()
+
+	var bodyString string
+	if caCertResponse.StatusCode == http.StatusOK {
```
[the remark this answers is not in the document]
Comment (cyBerta) [reply experiment/riseupvpn/riseupvpn_test.go]: fixed

File: experiment/riseupvpn/riseupvpn_test.go [hunk 11]
```diff
@@ -0,0 +1,480 @@
...
+	}
+	if *tk.ApiFailure != errorx.FailureEOFError {
+		t.Fatal("invalid ApiFailure")
+	}
+}
+
+func TestIntegrationFailureCaCertFetch(t *testing.T) {
+	measurer := riseupvpn.NewExperimentMeasurer(riseupvpn.Config{})
+	ctx, cancel := context.WithCancel(context.Background())
+	cancel()
```
[the remark this answers is not in the document]
Comment (cyBerta) [reply experiment/riseupvpn/riseupvpn_test.go:137]: fixed

File: experiment/riseupvpn/riseupvpn.go [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+
+func newGatewayConnection(tcpConnect archival.TCPConnectEntry, transportType string) *GatewayConnection {
+	return &GatewayConnection{
+		IP:            tcpConnect.IP,
+		Port:          tcpConnect.Port,
+		TransportType: transportType,
+	}
+}
+
+// AddCACertFetchTestKeys Adding generic ctx.Get() testKeys to riseupvpn specific test keys
```
[the remark this answers is not in the document]
Comment (cyBerta) [reply experiment/riseupvpn/riseupvpn.go]: I fixed the comment

File: experiment/riseupvpn/riseupvpn.go [hunk 13]
```diff
@@ -0,0 +1,279 @@
...
+	tk, err := caGetter.Get(ctx)
+
+	if err != nil {
+		log.Error("Getting CA cerificate failed. Aborting test.")
+		testkeys.AddCACertFetchTestKeys(tk)
+		measurement.TestKeys = testkeys
+		return nil
+	}
+
+	if ok := netx.CertPool.AppendCertsFromPEM([]byte(tk.HTTPResponseBody)); !ok {
```
[the remark this answers is not in the document]
Comment (cyBerta) [reply experiment/riseupvpn/riseupvpn.go]: I now create a local CertPool and append it to the urlgetter.Config.

**Comment by cyBerta** [comment]
I think we're getting closer. Both issues you mentioned - about the certpool and about refactoring the tests - have been resolved.

**Comment by bassosimone** [comment]
@cyBerta thank you! I've seen that [the tests failed](https://github.com/ooni/probe-engine/pull/978/checks?check_run_id=1355685856) with:

```
Error:     riseupvpn_test.go:83: unexpected FailingGateways value
```

Maybe a flaky hosts?

File: experiment/riseupvpn/riseupvpn_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (bassosimone) [reply experiment/riseupvpn/riseupvpn_test.go:273]: Wouldn't this hide issues with the riseupvpn backend that instead we want to know about?

**Review by bassosimone — APPROVED** [review]
Thank you @cyBerta for your excellent work in this PR! I have run tests locally and they are working correctly. I'm going to merge this PR later today!

🐳

**Comment by cyBerta** [comment]
Great! Thanks @bassosimone for your review. I'm happy to see the PR made it into probe-engine.

**Pull request merged**